From 7bdd60c52e6d60319bd45de0f7c3022d8d77a3d6 Mon Sep 17 00:00:00 2001
From: Rik van Riel <riel@redhat.com>
Date: Mon, 14 Jul 2014 19:19:06 -0400
Subject: [mm] numa: Remove BUG_ON() in __handle_mm_fault()

Message-id: <20140714151906.0455e267@annuminas.surriel.com>
Patchwork-id: 85802
O-Subject: [RHEL 7.1 PATCH] mm/numa: Remove BUG_ON() in __handle_mm_fault()
Bugzilla: 1119439
RH-Acked-by: Rafael Aquini <aquini@redhat.com>
RH-Acked-by: Larry Woodman <lwoodman@redhat.com>

Fixes bug 1119439

commit 107437febd495a50e2cd09c81bbaa84d30e57b07
Author: Rik van Riel <riel@redhat.com>
Date:   Tue Apr 29 15:36:15 2014 -0400

    mm/numa: Remove BUG_ON() in __handle_mm_fault()

    Changing PTEs and PMDs to pte_numa & pmd_numa is done with the
    mmap_sem held for reading, which means a pmd can be instantiated
    and turned into a numa one while __handle_mm_fault() is examining
    the value of old_pmd.

    If that happens, __handle_mm_fault() should just return and let
    the page fault retry, instead of throwing an oops. This is
    handled by the test for pmd_trans_huge(*pmd) below.

    Signed-off-by: Rik van Riel <riel@redhat.com>
    Reviewed-by: Naoya Horiguchi <n-horiguchi@ah.jp.nec.com>
    Reported-by: Sunil Pandey <sunil.k.pandey@intel.com>
    Signed-off-by: Peter Zijlstra <peterz@infradead.org>
    Cc: Andrew Morton <akpm@linux-foundation.org>
    Cc: Johannes Weiner <hannes@cmpxchg.org>
    Cc: Kirill A. Shutemov <kirill.shutemov@linux.intel.com>
    Cc: Linus Torvalds <torvalds@linux-foundation.org>
    Cc: Mel Gorman <mgorman@suse.de>
    Cc: linux-mm@kvack.org
    Cc: lwoodman@redhat.com
    Cc: dave.hansen@intel.com
    Link: http://lkml.kernel.org/r/20140429153615.2d72098e@annuminas.surriel.com
    Signed-off-by: Ingo Molnar <mingo@kernel.org>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/mm/memory.c b/mm/memory.c
index 9e33057..986ddb2 100644
--- a/mm/memory.c
+++ b/mm/memory.c
@@ -3761,9 +3761,6 @@ int handle_mm_fault(struct mm_struct *mm, struct vm_area_struct *vma,
   }
  }
 
- /* THP should already have been handled */
- BUG_ON(pmd_numa(*pmd));
-
  /*
   * Use __pte_alloc instead of pte_alloc_map, because we can't
   * run pte_offset_map on the pmd, if an huge pmd could
-- 
1.7.1